File: sdk/nodejs/runtime/log.ts
// Copyright 2016-2017, Pulumi Corporation.  All rights reserved.

import { getEngine, rpcKeepAlive } from "./settings";
let engproto = require("../proto/engine_pb.js");

// Log offers the ability to log messages in a way that integrate tightly with the resource engine's interface.
export class Log {
    private static errcnt = 0;
    private static lastLog: Promise<any> = Promise.resolve();

    // hasErrors returns true if any errors have occurred in the program.
    public static hasErrors(): boolean {
        return Log.errcnt > 0;
    }

    // debug logs a debug-level message that is generally hidden from end-users.
    public static debug(msg: string): void {
        let engine: Object | undefined = getEngine();
        if (engine) {
            Log.log(engine, engproto.LogSeverity.DEBUG, msg);
        }
        else {
            // ignore debug messages when no engine is available.
        }
    }

    // info logs an informational message that is generally printed to stdout during resource operations.
    public static info(msg: string): void {
        let engine: Object | undefined = getEngine();
        if (engine) {
            Log.log(engine, engproto.LogSeverity.INFO, msg);
        }
        else {
            console.log(`info: [runtime] ${msg}`);
        }
    }

    // warn logs a warning to indicate that something went wrong, but not catastrophically so.
    public static warn(msg: string): void {
        let engine: Object | undefined = getEngine();
        if (engine) {
            Log.log(engine, engproto.LogSeverity.WARNING, msg);
        }
        else {
            console.warn(`warning: [runtime] ${msg}`);
        }
    }

    // error logs a fatal error to indicate that the tool should stop processing resource operations immediately.
    public static error(msg: string): void {
        Log.errcnt++; // remember the error so we can suppress leaks.

        let engine: Object | undefined = getEngine();
        if (engine) {
            Log.log(engine, engproto.LogSeverity.ERROR, msg);
        }
        else {
            console.error(`error: [runtime] ${msg}`);
        }
    }

    private static log(engine: any, sev: any, msg: string): void {
        // Ensure we log everything in serial order.
        Log.lastLog = Log.lastLog.then(() => {
            return new Promise((resolve) => {
                let req = new engproto.LogRequest();
                req.setSeverity(sev);
                req.setMessage(msg);
                engine.log(req, () => { resolve(); });
            });
        });
    }
}

File: sdk/dotnet/Pulumi/Log.cs
// Copyright 2016-2019, Pulumi Corporation

using System;

namespace Pulumi
{
    /// <summary>
    /// Logging functions that can be called from a .NET application that will be logged to the
    /// <c>Pulumi</c> log stream.  These events will be printed in the terminal while the Pulumi app
    /// runs, and will be available from the Web console afterwards.
    /// </summary>
    public static class Log
    {
        /// <summary>
        /// Logs a debug-level message that is generally hidden from end-users.
        /// </summary>
        public static void Debug(string message, Resource? resource = null, int? streamId = null, bool? ephemeral = null)
            => Deployment.InternalInstance.Logger.DebugAsync(message, resource, streamId, ephemeral);

        /// <summary>
        /// Logs an informational message that is generally printed to stdout during resource
        /// operations.
        /// </summary>
        public static void Info(string message, Resource? resource = null, int? streamId = null, bool? ephemeral = null)
            => Deployment.InternalInstance.Logger.InfoAsync(message, resource, streamId, ephemeral);

        /// <summary>
        /// Logs a warning to indicate that something went wrong, but not catastrophically so.
        /// </summary>
        public static void Warn(string message, Resource? resource = null, int? streamId = null, bool? ephemeral = null)
            => Deployment.InternalInstance.Logger.WarnAsync(message, resource, streamId, ephemeral);

        /// <summary>
        /// Logs a fatal condition. Consider raising an exception
        /// after calling Error to stop the Pulumi program.
        /// </summary>
        public static void Error(string message, Resource? resource = null, int? streamId = null, bool? ephemeral = null)
            => Deployment.InternalInstance.Logger.ErrorAsync(message, resource, streamId, ephemeral);

        /// <summary>
        /// Logs an exception. Consider raising the exception after
        /// calling this method to stop the Pulumi program.
        /// </summary>
        public static void Exception(Exception exception, Resource? resource = null, int? streamId = null, bool? ephemeral = null)
            => Error(exception.ToString(), resource, streamId, ephemeral);
    }
}
